File: filter_changes.py
#!/usr/bin/env python3
"""
This script connects to an Ethereum node via WebSocket and queries filter changes/logs
until interrupted with Ctrl+C.
"""

import argparse
import asyncio
import logging
import signal
import sys

import rpc.common.websocket

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


async def main():
    """Main function to query the blocks via WebSocket."""

    # Parse command-line arguments
    parser = argparse.ArgumentParser(
        description="Connects to an Ethereum node via WebSocket and create a state change filter."
    )
    parser.add_argument(
        "websocket_url",
        type=str,
        nargs='?',  # Make the argument optional
        default="ws://127.0.0.1:8545",  # Set the default value
        help="The WebSocket URL of the Ethereum node (default: ws://127.0.0.1:8545)",
    )
    parser.add_argument(
        "ca_file",
        type=str,
        nargs='?',  # Make the argument optional
        default=None,  # Set the default value
        help="The path to your WSS server's *public* certificate file .pem or .crt (default: None)",
    )
    args = parser.parse_args()

    # Create the WebSocket client
    client = rpc.common.websocket.Client(args.websocket_url, args.ca_file)
    shutdown_event = asyncio.Event()

    # Setup signal handler for graceful shutdown
    async def signal_handler():
        print("")
        logger.info("🏁 Received interrupt signal (Ctrl+C)")
        shutdown_event.set()

    loop = asyncio.get_running_loop()
    for sig in [signal.SIGINT, signal.SIGTERM]:
        loop.add_signal_handler(sig, lambda: asyncio.create_task(signal_handler()))

    status = 0
    try:
        # Connect to Ethereum node
        await client.connect()
        logger.info(f"✅ Successfully connected to Ethereum node at {client.node_url}")

        # Register a new filter with some topic and periodically pull changes
        delay = 2
        f = await client.w3.eth.filter({"topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3e0"]})
        logger.info(f"✅ State change filter registered")
        while not shutdown_event.is_set():
            changes = await client.w3.eth.get_filter_changes(f.filter_id)
            if changes:
                logger.info(f"Changes: {changes}")
            else:
                logger.info(f"No change received")
            logs = await client.w3.eth.get_filter_logs(f.filter_id)
            if logs:
                logger.info(f"Logs: {logs}")
            else:
                logger.info(f"No log received")
            # Wait for some time before iterating
            await asyncio.sleep(delay)
        await client.w3.eth.uninstall_filter(f.filter_id)

        logger.info(f"✅ State change filter unregistered")
    except KeyboardInterrupt:
        logger.info("Interrupted by user")
    except Exception as e:
        logger.error(f"❌ Unexpected error: {e}")
        status = 1
    finally:
        # Cleanup
        await client.disconnect()
        sys.exit(status)


if __name__ == "__main__":
    """ Usage: python state_change_filter.py """
    asyncio.run(main())

File: internal/tools/filter_changes.go
package tools

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/erigontech/rpc-tests/internal/rpc"
	"github.com/urfave/cli/v2"
)

// ERC20 Transfer event topic
const transferTopic = "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3e0"

var filterChangesCommand = &cli.Command{
	Name:  "filter-changes",
	Usage: "Create ERC20 Transfer filter and poll changes/logs via WebSocket",
	Flags: []cli.Flag{
		&cli.StringFlag{
			Name:  "url",
			Value: "ws://127.0.0.1:8545",
			Usage: "WebSocket URL of the Ethereum node",
		},
	},
	Action: runFilterChanges,
}

func runFilterChanges(c *cli.Context) error {
	url := c.String("url")

	conn, err := rpc.Dial(url)
	if err != nil {
		return fmt.Errorf("connect to %s: %w", url, err)
	}
	defer conn.Close()
	log.Printf("Successfully connected to Ethereum node at %s", url)

	// Create filter with Transfer topic
	var filterResp jsonRPCResponse
	err = conn.CallJSON(jsonRPCRequest{
		Jsonrpc: "2.0",
		Method:  "eth_newFilter",
		Params:  []any{map[string]any{"topics": []string{transferTopic}}},
		ID:      1,
	}, &filterResp)
	if err != nil {
		return fmt.Errorf("create filter: %w", err)
	}
	if filterResp.Error != nil {
		return fmt.Errorf("create filter RPC error: %v", filterResp.Error)
	}
	filterID, ok := filterResp.Result.(string)
	if !ok {
		return fmt.Errorf("unexpected filter ID type: %T", filterResp.Result)
	}
	log.Printf("State change filter registered: %s", filterID)

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	delay := 2 * time.Second
	ticker := time.NewTicker(delay)
	defer ticker.Stop()

	for {
		// Get filter changes
		var changesResp jsonRPCResponse
		err = conn.CallJSON(jsonRPCRequest{
			Jsonrpc: "2.0",
			Method:  "eth_getFilterChanges",
			Params:  []any{filterID},
			ID:      2,
		}, &changesResp)
		if err != nil {
			log.Printf("Error getting filter changes: %v", err)
		} else if changes, ok := changesResp.Result.([]any); ok && len(changes) > 0 {
			log.Printf("Changes: %v", changes)
		} else {
			log.Printf("No change received")
		}

		// Get filter logs
		var logsResp jsonRPCResponse
		err = conn.CallJSON(jsonRPCRequest{
			Jsonrpc: "2.0",
			Method:  "eth_getFilterLogs",
			Params:  []any{filterID},
			ID:      3,
		}, &logsResp)
		if err != nil {
			log.Printf("Error getting filter logs: %v", err)
		} else if logs, ok := logsResp.Result.([]any); ok && len(logs) > 0 {
			log.Printf("Logs: %v", logs)
		} else {
			log.Printf("No log received")
		}

		select {
		case <-sigs:
			log.Printf("Received interrupt signal")
			// Uninstall filter
			var uninstallResp jsonRPCResponse
			_ = conn.CallJSON(jsonRPCRequest{
				Jsonrpc: "2.0",
				Method:  "eth_uninstallFilter",
				Params:  []any{filterID},
				ID:      4,
			}, &uninstallResp)
			log.Printf("State change filter unregistered")
			return nil
		case <-ticker.C:
		}
	}
}
